src/Middleware/MultipartParser.php

```diff
@@ -0,0 +1,266 @@
+<?php
+
+namespace React\Http\Middleware;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use React\Http\HttpBodyStream;
+use React\Http\UploadedFile;
+use RingCentral\Psr7;
+
+final class MultipartParser
+{
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var string
+     */
+    protected $ending = '';
+
+    /**
+     * @var int
+     */
+    protected $endingSize = 0;
+
+    /**
+     * @var string
+     */
+    protected $boundary;
+
+    /**
+     * @var RequestInterface
+     */
+    protected $request;
+
+    /**
+     * @var HttpBodyStream
+     */
+    protected $body;
+
+    /**
+     * @var callable
+     */
+    protected $onDataCallable;
+
+    public static function parseRequest(ServerRequestInterface $request)
+    {
+        $parser = new self($request);
+        return $parser->parse();
+    }
+
+    private function __construct(ServerRequestInterface $request)
+    {
+        $this->request = $request;
+    }
+
+    private function parse()
+    {
+        $this->buffer = (string)$this->request->getBody();
+
+        $this->determineStartMethod();
+
+        return $this->request;
+    }
+
+    private function determineStartMethod()
+    {
+        if (!$this->request->hasHeader('content-type')) {
+            $this->findBoundary();
+        }
+
+        $contentType = $this->request->getHeaderLine('content-type');
+        preg_match('/boundary="?(.*)"?$/', $contentType, $matches);
+        if (isset($matches[1])) {
+            $this->setBoundary($matches[1]);
+            $this->parseBuffer();
+        }
+
+        $this->findBoundary();
+    }
+
+    private function setBoundary($boundary)
+    {
+        $this->boundary = $boundary;
+        $this->ending = $this->boundary . "--\r\n";
+        $this->endingSize = strlen($this->ending);
+    }
+
+    private function findBoundary()
+    {
+        if (substr($this->buffer, 0, 3) === '---' && strpos($this->buffer, "\r\n") !== false) {
+            $boundary = substr($this->buffer, 2, strpos($this->buffer, "\r\n"));
+            $boundary = substr($boundary, 0, -2);
+            $this->setBoundary($boundary);
+            $this->parseBuffer();
+        }
+    }
+
+    private function parseBuffer()
+    {
+        $chunks = preg_split('/-+' . $this->boundary . '/', $this->buffer);
```

Comment:
The boundary may potentially contain characters that would be interpreted as part of the regex here

Reply:
Good catch, I'll fortify it 👍

Reply:
Actually I've swapped it out with an explode for simplicity